stdlib/domain.ml

```diff
@@ -104,40 +125,75 @@ module DLS = struct
         if idx < s then s else compute_new_size (2 * s)
       in
       let new_sz = compute_new_size sz in
-      let new_st = Array.make new_sz unique_value in
+      let new_st = Array.make new_sz Obj_opt.none in
       Array.blit st 0 new_st 0 sz;
-      set_dls_state new_st;
-      new_st
+      (* We want a implementation that is safe with respect to
+         single-domain multi-threading: retry if the DLS state has
+         changed under our feet.
+         Note that the number of retries will be very small in
+         contended scenarios, as the array only grows, with
+         exponential resizing. *)
+      if compare_and_set_dls_state st new_st
+      then new_st
+      else maybe_grow idx
     end
 
-  let set (idx, _init) x =
+  let set (type a) (idx, _init) (x : a) =
     let st = maybe_grow idx in
     (* [Sys.opaque_identity] ensures that flambda does not look at the type of
      * [x], which may be a [float] and conclude that the [st] is a float array.
      * We do not want OCaml's float array optimisation kicking in here. *)
-    st.(idx) <- Obj.repr (Sys.opaque_identity x)
+    st.(idx) <- Obj_opt.some (Sys.opaque_identity x)
+
 
-  let get (idx, init) =
+  let[@inline never] array_compare_and_set a i oldval newval =
+    (* Note: we cannot use [@poll error] due to the
+       allocations on a.(i) in the Double_array case. *)
+    let curval = a.(i) in
+    if curval == oldval then (
+      Array.unsafe_set a i newval;
+      true
+    ) else false
+
+  let get (type a) ((idx, init) : a key) : a =
     let st = maybe_grow idx in
-    let v = st.(idx) in
-    if v == unique_value then
-      let v' = Obj.repr (init ()) in
-      st.(idx) <- (Sys.opaque_identity v');
-      Obj.magic v'
-    else Obj.magic v
-
-  let get_initial_keys () : (int * Obj.t) list =
+    let obj = st.(idx) in
+    if Obj_opt.is_some obj
+    then (Obj_opt.unsafe_get obj : a)
+    else begin
+      let v : a = init () in
+      let new_obj = Obj_opt.some (Sys.opaque_identity v) in
+      (* At this point, [st] or [st.(idx)] may have been changed
+         by another thread on the same domain.
+
+         If [st] changed, it was resized into a larger value,
+         we can just reuse the new value.
+
+         If [st.(idx)] changed, we drop the current value to avoid
+         letting other threads observe a 'revert' that forgets
+         previous modifications. *)
+      let st = get_dls_state () in
+      if array_compare_and_set st idx obj new_obj
+      then v
+      else begin
+        (* if st.(idx) changed, someone must have initialized
+           the key in the meantime. *)
+        let updated_obj = st.(idx) in
+        if Obj_opt.is_some updated_obj
+        then (Obj_opt.unsafe_get updated_obj : a)
+        else assert false
```

Comment:
Nit: saves one line.
```suggestion
        assert (Obj_opt.is_some updated_obj);
        (Obj_opt.unsafe_get updated_obj : a)
```

Reply:
I used the current version to remain statically safe under the `-noassert` flag, which disables all `assert` expressions except `assert false`.

Reply:
You can still save one line:
```suggestion
        if not (Obj_opt.is_some updated_obj) then assert false;
        (Obj_opt.unsafe_get updated_obj : a)
```
Don't you love double negations?